trino/trino.go

```diff
@@ -976,11 +1022,34 @@ func (qr *driverRows) fetch(allowEOF bool) error {
 	qr.rowindex = 0
 	qr.data = qresp.Data
 	qr.nextURI = qresp.NextURI
+
+	qrStats := QueryProgressCallbackResponse{
+		ID:         qresp.ID,
+		QueryStats: qresp.Stats,
+		Error:      qresp.Error,
+	}
+
 	if len(qr.data) == 0 {
 		if qr.nextURI != "" {
+			if qr.stmt.conn.callback != nil {
+				if qrStats.QueryStats.State == "RUNNING" {
+					currentTime := time.Now()
+					diff := currentTime.Sub(qr.stmt.conn.callbackPeriod.LastCallbackTime).Milliseconds()
+					period := qr.stmt.conn.callbackPeriod.Period.Milliseconds()
+					// Check if period has passed
+					if diff >= period {
+						qr.stmt.conn.callback.ProgressUpdated(qrStats)
```

Comment:
Can this be blocking? I think instead of calling it here, you should write to a (buffered?) channel and have a goroutine reading from it and calling `ProgressUpdated`. If the channel is full, maybe it would be ok to drop updates.

Reply:
Hmm, I haven't had a thought on this that it could be blocking (depends on the ProgressUpdated that user defines, and period that user set ?)
Thx for insight. i'll try it out using channel.

Reply:
If you'll have any issues implementing it I could help, we can sync on the Trino Slack.

Reply:
I greatly appreciate on giving detailed feedback and pointing out places where it could lead to problems. I am learning a lot :)

I'll first try out on using channel and share to code with the rest of changes I made.

Reply:
What do you think of this change?
https://github.com/trinodb/trino-go-client/pull/14/commits/fa70b40b593afc0c49377687be1e667a70ccb79a

I wonder if it is allright to call go routine like this (simply open channel and send data if conn has callback)
```go
c := make(chan QueryProgressCallbackResponse)
go qr.updateCallback(c)
	if len(qr.data) == 0 {
		if qr.stmt.conn.callback != nil {
			c <- qrStats
		}
		if qr.nextURI != "" {
			return qr.fetch(allowEOF)
		}
		if allowEOF {
			qr.err = io.EOF
			return qr.err
		}
	}
	...
	qr.rowsAffected = qresp.UpdateCount
	if qr.stmt.conn.callback != nil {
		c <- qrStats
	}
```

Reply:
You'll need to:
1. Create two unbuffered channels when the query begins in `exec()`, one for query stats and another one to signal when the query is done. The second one should be of an empty type `struct{}`, because it won't transmit data, only be closed.
2. Start a goroutine that would read from the channel and call the update callback. It's important to control when this goroutine should end (when the query is done). That's why we need the second channel.
3. When receiving a query response, send query stats into the channel. Ignore the error when the channel is full.
4. When the query finishes, close the stop channel and the stats channel (in this order).

An example goroutine:
```go
done := make(chan struct{})
go func() {
  for {
    select {
    case stats := <-statsCh:
      update(stats)
    case <-done:
      return
    }
  }
}()
```

Sending to a channel but ignoring when it's full:
```go
select {
    case statsCh <- stats:
    default:
        // ignore when can't send stats
    }
```

Reply:
Quick question, why channel at exec and not at fetch? I thought exec is only called once, so need to worry for blocking.

Reply:
You should do the setup once, in exec, when the query starts. So create all required channels and set up the goroutine once, instead of creating goroutines for every HTTP request. This will help avoid issues with memory usage. It's equally important to ensure the goroutine will finish when the query is done, to avoid leaking it.

Reply:
I tried to do as you said, but how do I call statsCh inside fetch function? I can't pass it as parameter so statsCh variable does not exist in fetch function scope.

Reply:
Add fields to the statement and driverRows structs.